### ISSUE TITLE
Key becomes stuck if a Dualshock 4 trigger is mapped to it

### ISSUE THREAD

**Issue by maxigaz** [dosbox-staging/dosbox-staging#943]
**Describe the bug**
In Realms of Chaos, when I map either trigger buttons of my DualShock 4 to SHIFT (used for looking up/down when held down and up/down is pressed simultaneously), the game behaves as if SHIFT is held down all the time. This makes the pause menu and the main menu unusuble and crouching/aiming up in a game is not possible either.

The issue with the menu can be temporarily fixed by pressing the trigger again, but it has to be repeated each time after a menu item is selected or the menu is brought up with Esc.

**To Reproduce**
1. Launch the game with the attached config file and mapper file (or press Ctrl-F1 and manually map RSHIFT to LT or RT of a DualShock 4).
2. Menus can’t be used until LT or RT is pressed.
3. During a game, pressing down/up is interpreted as looking down/up instead of crouching/aiming upwards.

**Expected behaviour**
Holding down SHIFT is only simulated when the trigger button is actually held.

**Environment (please complete the following information):**
- OS and version: Manjaro Linux x86-64
- Relevant hardware info: Relevant hardware info: Dual Core Intel Celeron G1610, Radeon HD 6670 with Mesa 20.3.4, 4 GB of RAM

**Version of dosbox-staging:**
v0.77.0-alpha-671-g49148e60

**Configuration parameters:**
See the attached `dosbox-staging.conf` and `mapper-0.76.sdl2.map`: [roc-conf.zip](https://github.com/dosbox-staging/dosbox-staging/files/6140976/roc-conf.zip)

**Additional context**
In the mapper window, LT is reported as “PS4 Controller Axis 4-” and RT as “PS4 Controller Axis 5-”.

**Comment by kcgen** [comment]
Thanks for another nice report @maxigaz!
Can you see if branch [`kc/6axis-triggers-1`](https://github.com/dosbox-staging/dosbox-staging/pull/944) (PR-944) fixes it? 

I don't have a Dualshock 4, so haven't been able confirm on my side.  Unfortunately I think to test it, you will need to scrub your mapping file and start fresh. 

If anyone else with a controller and some mapping file(s) can confirm that all is still OK (regression free), your help would be appreciated!

**Comment by maxigaz** [comment]
> Can you see if branch [`kc/6axis-triggers-1`](https://github.com/dosbox-staging/dosbox-staging/pull/944) (PR-944) fixes it?

Yes, I've just tried it and it works perfectly now, both with a Dualshock 4 and an XBox 360 controller. Thank you very much!

> Unfortunately you will need to scrub your mapping file and start fresh (to test this fix).

So, if want to keep using this build for a while, will I need to replace all the other mapping files as well, even if they don’t include mappings for any trigger buttons? (It’s not a big deal, I’m just wondering.)

**Comment by kcgen** [comment]
That's great news @maxigaz - thanks!

> will I need to replace all the other mapping files as well, even if they don’t include mappings for any trigger buttons? (It wouldn't be a big deal, I’m just wondering.)

Thanks for double checking; good news here - mapping files _that don't map the trigger buttons_  should be fine to keep using.

**Comment by maxigaz** [comment]
Sorry to revive this issue, but I’ve just realised that I’d only tested the right trigger and I’d completely forgotten about the left one.

Now, even with a fresh mapper file, LT still behaves the same as described in OP. Also, in the mapper window, it’s described as “PS4 Controller Axis 4-”.

I’m using the same build from branch `kc/6axis-triggers-1` because judging from the latest commit messages in `master`, I don’t think anything has been changed in this regard in the past few days.

Again, I apologise for not letting you know sooner.

**Comment by kcgen** [comment]
Thanks @maxigaz!

The change is part of the `master` branch. 

I had tested mapping my Logitech F710's triggers to left and right shift keys, respectively, but suspect there's a logic bug left.

**Comment by kcgen** [comment]
Hi @maxigaz ,

It "works fine" for me, but maybe the F710's left trigger differs from the Dual Shock's.
Here's what I did:
1. Delete current mapper file
2. Start current build of `master` branch and open the Mapper
3. Mouse-click on "Left-Shift" graphical keyboard button (now green)
4. Mouse-click on "Remove Mapping" (once) and "Remove Mapping" (twice) - so all of it's current mappings are gone.
5. Mouse-click on "Add Bind" followed by pulling the Left-Trigger on my F710 joystick. 
6. Repeat 3 through 5 for the "Right-Shift" keyboard button and Right-Trigger pairing.
7. Mouse-click save, exit, then quit DOSBox back to the host.

My mapper file contains these left and right shift mappings:

![2021-03-24_06-04](https://user-images.githubusercontent.com/1557255/112315049-f29f2180-8c66-11eb-9f62-7bf3f90c91a8.png)

Can you try the same (using the current `master` branch), and post your two mappings?

**Comment by maxigaz** [comment]
I've repeated the exact same steps using `master`. Here are some differences I noticed:

> 4. Mouse-click on "Remove Mapping" (once) and "Remove Mapping" (twice) - so all of it's current mappings are gone.

I clicked on “Remove Mapping” once and then it disappeared (so I couldn't click on it the second time) and Left Shift greyed out. Same with Right Shift.

> 9. Type some keyboard characters and confirm they're lowercase.

For me, they’re all in uppercase unless I pull the left trigger.

> When you have a chance, can you try the same (using the current master branch), and post your two mappings?

```
key_lshift "stick_0 axis 4 0" 
key_rshift "stick_0 axis 5 1" 
```

**Comment by kcgen** [comment]
Great digging @maxigaz!

> Edit: For good measure, I also checked my XBox 360 Controller. With that, everything is exactly the same as you described

That explains it!  So we've got:
 - two 6-axis controller where left-trigger is "axis 2"
 - the Dual Shock 6-axis controller where left-trigger is "axis 4"
 - All three the right-trigger is "axis 5".

Axis 4 is the right analog stick on the F710 and likely the XBox 360 - so we hit a conflict here that will cause a regression on that stick control.

I think because of this conflict, we can't proceed.  We need to upgrade from the "dumb" code-based joystick API to the new Controller-based API where SDL always gives us the same code for "left trigger".

Fortunately the fix for you is easy, in the meantime. Whenever you want to map your Dual Shock's left-trigger, edit the mapper file and change the trailing 0 in `axis 4 0` to a `1`. So in your left-shift example:

```
key_lshift "stick_0 axis 4 1" 
```

Because this is a work around, let's keep this issue open.

**Comment by maxigaz** [comment]
> In the meantime, whenever you want to map your Dual Shock's left-trigger, edit the mapper file and change the trailing 0 in `axis 4 0` to a `1`. So in your left-shift example:
> 
> ```
> key_lshift "stick_0 axis 4 1" 
> ```

I can confirm that this is a good temporary fix. Thank you!
